### PR TITLE
 Kt-20305 - Inspection: Refactor sealed sub-class to object

### PR DESCRIPTION
**Pull request by Megamiun**
#KT-20305 Fixed(Continuation of #1361, sorry for the delay)

https://youtrack.jetbrains.com/issue/KT-20305

### CODE REVIEW

File: idea/src/org/jetbrains/kotlin/idea/quickfix/sealedSubClassToObject/ConvertSealedSubClassToObjectFix.kt [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+        modifierList?.getModifier(KtTokens.OPEN_KEYWORD)?.delete()
+    }
+
+    private fun KtClass.changeToObject(factory: KtPsiFactory) {
+        getClassOrInterfaceKeyword()?.replace(factory.createExpression(KtTokens.OBJECT_KEYWORD.value))
+        primaryConstructor?.delete()
+    }
+
+    private fun KtClass.transformToObject(factory: KtPsiFactory) {
+        replace(factory.createDeclaration<KtObjectDeclaration>(text))
```
Comment (mglukhikh) [inline idea/src/org/jetbrains/kotlin/idea/quickfix/sealedSubClassToObject/ConvertSealedSubClassToObjectFix.kt]: There is already `factory.createObject`

File: idea/src/org/jetbrains/kotlin/idea/quickfix/sealedSubClassToObject/ConvertSealedSubClassToObjectFix.kt [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+        element.transformToObject(factory)
+    }
+
+    private fun KtClass.removeOpen() {
+        modifierList?.getModifier(KtTokens.OPEN_KEYWORD)?.delete()
+    }
+
+    private fun KtClass.changeToObject(factory: KtPsiFactory) {
+        getClassOrInterfaceKeyword()?.replace(factory.createExpression(KtTokens.OBJECT_KEYWORD.value))
+        primaryConstructor?.delete()
```
Comment (mglukhikh) [inline idea/src/org/jetbrains/kotlin/idea/quickfix/sealedSubClassToObject/ConvertSealedSubClassToObjectFix.kt]: You should also delete all secondary constructors, if any.

File: idea/src/org/jetbrains/kotlin/idea/quickfix/sealedSubClassToObject/ConvertSealedSubClassToObjectFix.kt [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+    private fun changeDeclaration(element: KtClass) {
+        val factory = KtPsiFactory(element)
+
+        element.removeOpen()
+        element.changeToObject(factory)
+        element.transformToObject(factory)
+    }
+
+    private fun KtClass.removeOpen() {
+        modifierList?.getModifier(KtTokens.OPEN_KEYWORD)?.delete()
```
Comment (mglukhikh) [inline idea/src/org/jetbrains/kotlin/idea/quickfix/sealedSubClassToObject/ConvertSealedSubClassToObjectFix.kt]: Here please use `addModifier(FINAL_KEYWORD)` directly on `KtClass`. This function includes necessary logic to delete incompatible modifiers (like open, abstract or sealed) and not to add redundant modifiers (like final in most cases).

File: idea/src/org/jetbrains/kotlin/idea/inspections/CanSealedSubClassBeObject.kt [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+class CanSealedSubClassBeObject : AbstractKotlinInspection(), CleanupLocalInspectionTool {
+    override fun buildVisitor(holder: ProblemsHolder, isOnTheFly: Boolean, session: LocalInspectionToolSession): PsiElementVisitor {
+        return object : KtVisitorVoid() {
+            override fun visitClass(klass: KtClass) {
+                if (!klass.isSealed()) return
+
+                klass.getSubclasses()
+                    .withEmptyConstructors()
+                    .thatAreFinal()
+                    .forEach { reportPossibleObject(it) }
```
Comment (mglukhikh) [inline idea/src/org/jetbrains/kotlin/idea/inspections/CanSealedSubClassBeObject.kt]: There are some more rules that should be either considered here (to remove suggestions for these cases, it's easier) or processed inside fix (to convert them to object correctly). In particular: 1) Class, but not object, may have companion object(s) inside; 2) Class, but not object, may have inner class(es) inside.

File: idea/src/org/jetbrains/kotlin/idea/quickfix/sealedSubClassToObject/ConvertSealedSubClassToObjectFix.kt [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+
+    /**
+     * Replace Kotlin instantiations to a straightforward call to the singleton.
+     */
+    private fun Map<Language, List<PsiElement>>.replaceKotlin(klass: KtClass) {
+        val list = this[KOTLIN_LANG] ?: return
+        val singletonCall = KtPsiFactory(klass).buildExpression { appendName(klass.nameAsSafeName) }
+
+        list.filter { it.node.elementType == KtNodeTypes.CALL_EXPRESSION }
+            .forEach { it.replace(singletonCall) }
```
Comment (mglukhikh) [inline idea/src/org/jetbrains/kotlin/idea/quickfix/sealedSubClassToObject/ConvertSealedSubClassToObjectFix.kt]: It's probably correct, but we should add some tests for: 1) Callable reference usages like `MyClass::myFunction` or `MyClass()::myFunction`; 2) Nested class usages liked `MyClass::NestedClass()` or `MyClass()::NestedClass()`.

File: idea/src/org/jetbrains/kotlin/idea/inspections/CanSealedSubClassBeObject.kt [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+import org.jetbrains.kotlin.idea.search.declarationsSearch.searchInheritors
+import org.jetbrains.kotlin.lexer.KtTokens
+import org.jetbrains.kotlin.psi.KtClass
+import org.jetbrains.kotlin.psi.KtVisitorVoid
+
+class CanSealedSubClassBeObject : AbstractKotlinInspection(), CleanupLocalInspectionTool {
+    override fun buildVisitor(holder: ProblemsHolder, isOnTheFly: Boolean, session: LocalInspectionToolSession): PsiElementVisitor {
+        return object : KtVisitorVoid() {
+            override fun visitClass(klass: KtClass) {
+                if (!klass.isSealed()) return
```
Comment (mglukhikh) [inline idea/src/org/jetbrains/kotlin/idea/inspections/CanSealedSubClassBeObject.kt]: You should use `KtDeclaration.getModalityFromDescriptor()` to get declaration modality (see detailed comment below)

**Comment by Megamiun** [comment]
You can leave the changes to me, @mglukhikh. No worries. Probably it is best this way, so I can get used more to the IDEA particularities.

**Comment by mglukhikh** [comment]
@Megamiun Ok

File: idea/src/org/jetbrains/kotlin/idea/inspections/CanSealedSubClassBeObject.kt [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+                    .thatAreFinal()
+                    .forEach { reportPossibleObject(it) }
+            }
+
+            private fun reportPossibleObject(klass: KtClass) {
+                val keyword = klass.getClassOrInterfaceKeyword() ?: return
+                holder.registerProblem(
+                    keyword,
+                    "Sealed Sub-class can be changed To Object",
+                    ProblemHighlightType.INFORMATION,
```
[the remark this answers is not in the document]
Comment (Megamiun) [reply idea/src/org/jetbrains/kotlin/idea/inspections/CanSealedSubClassBeObject.kt]: If I don't use INFORMATION,  the tests go red saying that only INFORMATION problems can have the word `can` inside their descriptions. Changed to `should`.

**Comment by mglukhikh** [comment]
@Megamiun I reviewed your inspection again and mostly it's OK. I will recheck it on real code base and merge. Thank you.

**Comment by mglukhikh** [comment]
NB: in future, please do not set `cleanupTool` to `true` unless your are absolutely sure your inspection is applicable as part of automatic code cleanup. Same applies to inheritance from ` CleanupLocalInspectionTool`.

**Pull request closed without merging**